pkg/mimirpb/prealloc_rw2.go

```diff
@@ -30,27 +34,50 @@ func FromWriteRequestToRW2Request(rw1 *WriteRequest, commonSymbols []string, off
 	defer reuseSymbolsTable(symbols)
 	symbols.ConfigureCommonSymbols(offset, commonSymbols)
 
+	var metadataMap map[string]*MetricMetadata
+	if len(rw1.Metadata) > 0 {
+		metadataMap = make(map[string]*MetricMetadata, len(rw1.Metadata))
+		for _, meta := range rw1.Metadata {
+			metadataMap[meta.MetricFamilyName] = meta
+		}
+	}
+
 	rw2Timeseries := make([]TimeSeriesRW2, 0, len(rw1.Timeseries)+len(rw1.Metadata)) // TODO: Pool-ify this allocation
 	for _, ts := range rw1.Timeseries {
 		refs := make([]uint32, 0, len(ts.Labels)*2) // TODO: Pool-ify this allocation
+		metricName := ""
 		for i := range ts.Labels {
+			if ts.Labels[i].Name == labels.MetricName {
+				metricName = ts.Labels[i].Value
+			}
 			refs = append(refs, symbols.Symbolize(ts.Labels[i].Name), symbols.Symbolize(ts.Labels[i].Value))
 		}
 
+		var metadata MetadataRW2
+		if len(rw1.Metadata) > 0 {
+			if rw1Metadata, ok := metadataMap[metricName]; ok {
+				metadata = FromMetricMetadataToMetadataRW2(rw1Metadata, symbols)
+			}
+			delete(metadataMap, metricName)
+		}
+
 		rw2Timeseries = append(rw2Timeseries, TimeSeriesRW2{
 			LabelsRefs:       refs,
 			Samples:          ts.Samples,
 			Histograms:       ts.Histograms,
 			Exemplars:        FromExemplarsToExemplarsRW2(ts.Exemplars, symbols),
-			Metadata:         MetadataRW2{},
+			Metadata:         metadata,
 			CreatedTimestamp: ts.CreatedTimestamp,
 		})
 	}
 
 	// Represent metadata as extra, empty timeseries rather than attaching it to existing series.
 	// We never replicate metadata if there are multiple series with the same name, and it removes the requirement to match up the metadata to the right series.
```

Comment:
[nit] Can you update this comment, to explain that here we append the remaining metadata belonging to series that didn't have any sample?